### ISSUE TITLE
Performance is very slow.

### ISSUE THREAD

**Issue by junderw** [junderw/tiny-secp256k1-wasm#4]
I used the following script from the root directory (I have an installed instance of tiny-secp256k1 in the folder next to this project)

The Bench results are not good:
13x longer than native JS
281x longer than native C++ bindings

I think the glue code generated by wasm-bindgen is slowing it down. Managing the wasm memory is very expensive.

```
$ node bench.js 
Staring bench for NATIVE 1000 times
Finished 1000 iterations of NATIVE in 51 ms
Staring bench for JS 1000 times
Finished 1000 iterations of JS in 1073 ms
Staring bench for WASM 1000 times
Finished 1000 iterations of WASM in 14356 ms
```
```js
const { sign } = require('./pkg');
const ecc = require('../tiny-secp256k1/js');
const eccNative = require('../tiny-secp256k1/native');
const crypto = require('crypto');

const ITER = 1e3;

function performBench(_sign, name, iter) {
  console.log(`Staring bench for ${name} ${iter} times`);
  const key = crypto.randomBytes(32);
  const hash = crypto.randomBytes(32);
  const before = Date.now();
  for(let i = 0; i < iter; i++) {
    _sign(hash, key)
  }
  const after = Date.now();
  console.log(`Finished ${iter} iterations of ${name} in ${after - before} ms`);
}

function main(args) {
  const iterations = args[0] || ITER;
  performBench(eccNative.sign, 'NATIVE', iterations);
  performBench(ecc.sign, 'JS', iterations);
  performBench(sign, 'WASM', iterations);
}

main(process.argv.slice(2))
```

#2

**Comment by yuki-js** [comment]
I think so too. The glue code by bindgen is not so fast and flexible. The original glue code also solves #3 .

**Comment by yuki-js** [comment]
But there might be some optimization workaround for bindgen.

**Comment by junderw** [comment]
from discord:

> @junderw I don't see anything in fn sign that looks like it should be a performance problem, assuming the underlying library you're using isn't really slow for some reason. From the issue you linked, the "13x slower than JS and 281x slower than C++" part sounds like maybe you didn't compile with optimizations enabled.
>
> How are you compiling the code? If you are using cargo build you need to pass it the --release flag in order to build with optimizations.
>
> I would also suggest trying the following cargo profile:

```
[profile.release]
codegen-units = 1
lto = true
# could also try opt-level = 3 if output size isn't an issue
opt-level = 'z'
panic = 'abort'
```

**Comment by junderw** [comment]
Resolved in c0fbcac5435653d84952aeb5ca830d481eed107b
